docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -201,44 +201,117 @@ The ShootState synchronization controller will become part of the gardenlet. It
+
+The safety mechanism is based on "cluster leases". Essentially, extension controllers must check if the seed has a valid "lease" for the shoot before attempting a reconciliation. The lease is just an additional timestamp field in the `Cluster` resource:
+
+```go
+// ClusterSpec is the spec for a Cluster resource.
+type ClusterSpec struct {
+	...
+	// LeaseExpiration indicates when the lease of this cluster will expire
+	LeaseExpiration metav1.MicroTime `json:"leaseExpiration"`
+}
+```
 
-When case 3 happens it means that the shoot has been migrated and the backuprestore container in the  __Destination Seed__ has started using the Shoot's backup entry. The backuprestore on the __Source Seed__ should be configured to shut itself and the etcd container down once it sees the newer timestamp. Shutting down the etcd will cause the kubernetes control plane components to go into crashloop backoff and the MCM will not be able to do anything as it will not be able to list nodes (this has to be verified with MCM).
+On a regular basis, e.g. every 2 minutes, and at each shoot reconciliation, `gardenlet` fetches the shoot from the Garden `kube-apiserver`, and if successful and the control plane migration hasn't been triggered it updates the `Cluster` resource with a renewed timestamp, e.g. 2 minutes from now. The extension controllers fetch this resource before each reconciliation and check if the lease has expired. If it hasn't, they reconcile as usual. If the lease has expired, they skip the reconciliation, since either the control plane migration has been triggered or `gardenlet` cannot reliably determine if the seed is still the owner.
 
-For this approach to work backups must be enabled for the Shoot which is migrated. Additionally, synchronization based on the timestamps in the backup entry depend on the frequency of backups made by the backuprestore container.
+**Note:** The `dns-external` extension controller is the only extension controller that neither needs the shoot's `kube-apiserver`, nor uses the `Cluster` resource for anything. Therefore, this controller will continue reconciling `DNSEntry` resources even after the source seed has lost the ownership of the shoot. In the PoC, we manually deleted the `DNSOwner` resources from the source seed cluster to prevent this from happening. Eventually, the `dns-external` controller should either use the "cluster leases" mechanism described here, or a different mechanism to ensure that it disables itself after the seed has lost the ownership of the shoot. 
```

Comment:
Well, what should I say... see https://github.com/gardener/external-dns-management/issues/161#issuecomment-806486232 and the following comments.

Reply:
/cc @mandelsoft to comment on this.